### ISSUE TITLE
autograd.Function that returns input as-is is silently incorrect with torch.compile

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#102970]
### 🐛 Describe the bug

```py
import torch

class Foo(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x

    @staticmethod
    def backward(ctx, gx):
        return gx * 0.5

x = torch.randn([], requires_grad=True)

def f(x):
    return Foo.apply(x)

y = torch.compile(f)(x)
result, = torch.autograd.grad(y, x)
print(result)  # 1.0, but should be 0.5
```

This is a relatively common case, used in situations where people want to scale their gradients. I dug into this a while ago, and it looked like AOTAutograd got confused about the output of Foo.

### Versions

main

cc @ezyang @gchanan @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @Chillee @samdow @kshitij12345 @janeyx99

**Comment by janeyx99** [comment]
Tentatively marking as hipri cuz silent incorrectness

**Comment by albanD** [comment]
> If that's the case- how does autograd.Function handle this situation? Does it have some special handling for when a user's custom autograd.Function directly returns one of the inputs? 

Yes it does
 https://github.com/pytorch/pytorch/blob/f1f57e1e54593b279dcb856cc9fd8d3cfcd4e44b/torch/csrc/autograd/custom_function.cpp#LL298C11-L298C11

**Comment by zou3519** [comment]
Here's a repro without Dynamo: https://gist.github.com/zou3519/1792e4cbce58b7b0819e1418045f797b , so something is likely going wrong where tracing is not replicating the eager-mode semantics

**Comment by bdhirsh** [comment]
It looks like this is happening because AOTAutograd has been making an assumption that is no longer valid:

- When AOTAutograd traces a graph where it gets an output that aliases an input (or is an input), it doesn't ahead-of-time trace out a backwards graph for that output. Instead, it uses "view-replay": in a runtime epilogue after running the compiled graph, it regenerates any outputs that are views as inputs, and directly returns them. There were a few reasons that we did this: the original one that I remember is that if you attempted to mutate the tensor afterwards, the autograd engine would yell at you: it sees a view getting mutated that has a custom autograd backward, which is not allowed. In particular, this would cause models that passed in eager (because they had ordinary view ops) to break in torch.compile (because they have special view ops that the autograd engine thinks has a custom backward).

- We shouldn't apply this treatment to outputs-that-alias-inputs that come from custom autograd functions, because we **can't** do view-replay on them afterwards - we need their backwards to be the actual custom backward that the user specified.


We could have dynamo graph break on any autograd.Functions where it detects outputs that alias inputs. Alternatively, we could figure out how to get AOTAutograd to recognize that a view has a custom backward, and not attempt to do the view-replay on it that we usually do.

Maybe we can just check the `grad_fn` on the view, and see if it is a custom grad_fn. I'll try that first.
